drivers/modem/hl7800.c

```diff
@@ -1829,17 +1829,40 @@ static void dns_work_cb(struct k_work *work)
 #if defined(CONFIG_DNS_RESOLVER) && !defined(CONFIG_DNS_SERVER_IP_ADDRESSES)
 	int ret;
 	struct dns_resolve_context *dnsCtx;
-	static const char *const dns_servers_str[] = { ictx.dns_v4_string,
+	struct sockaddr temp_addr;
+	bool valid_address = false;
+	static const char *const dns_servers_str[] = {
 #ifdef CONFIG_NET_IPV6
-						       ictx.dns_v6_string,
+		ictx.dns_v6_string,
 #endif
-						       NULL };
+#ifdef CONFIG_NET_IPV4
+		ictx.dns_v4_string,
+#endif
+		NULL};
+
+#ifdef CONFIG_NET_IPV6
+	if (IS_ENABLED(CONFIG_NET_IPV6)) {
+		valid_address = net_ipaddr_parse(ictx.dns_v6_string, strlen(ictx.dns_v6_string),
+						 &temp_addr);
+		if (!valid_address && IS_ENABLED(CONFIG_NET_IPV4)) {
+			/* IPv6 DNS string is not valid, replace it with IPv4 address and recheck */
+			strncpy(ictx.dns_v6_string, ictx.dns_v4_string, strlen(ictx.dns_v4_string));
+			valid_address = net_ipaddr_parse(ictx.dns_v6_string,
+							 strlen(ictx.dns_v6_string), &temp_addr);
+		}
+	} else
+#endif
+	{
```

Comment:
So IPv4 is always enabled or why there is no IPv4 check here? Asking because you have a proper check in added code above.